### ISSUE TITLE
Adding a prefab with BepInEx which can be instantiated by Photon?

### ISSUE THREAD

**Issue by StupidRepo** [BepInEx/HarmonyX#119]
Sorry if this issue doesn't fit here, I'm not really sure where to put it as it is kind of a BepInEx question and also a Photon Unity Networking (PUN) question.

Effectively, I'm trying to make an API for a multiplayer game where I can add items to the game and sync them across the network however (iirc) a prefab will need to be put in a specific folder for PUN (`Resources`, iirc?).

So, is it possible to make my own prefab instantiable by my plugin using PUN to sync it to others with the mod? I know I'll need an AssetBundle to hold the prefab but then I'm not really sure where to go from there. Other games I know with BepInEx custom item plugins don't use PUN (e.g. Lethal Company) so I wouldn't know any open source plugins I can 'borrow' code from...

**Comment by ManlyMarco** [comment]
You have to do it the same way as you would when making the game itself, have all people run the same code and resources so that the environments are all the same.

**Comment by StupidRepo** [comment]
What do you mean by that? Because what I'm saying is that the prefab itself needs to be physically in a folder in the Unity project, but as a simple modder with no access to the project then how can I add a prefab and register it to allow Photon not to complain?

Or am I being really stupid here? 😭

**Comment by ManlyMarco** [comment]
Just put it in an asset bundle and load it that way.

**Comment by StupidRepo** [comment]
But I thought PUN would only load things if it was in that special folder?

**Comment by ManlyMarco** [comment]
You can load ABs from wherever you want https://docs.unity3d.com/560/Documentation/ScriptReference/AssetBundle.LoadFromFile.html

**Comment by StupidRepo** [comment]
Yeah I know that but I'm pretty sure PUN has a weird restriction on networked prefabs. Maybe I'm wrong...

**Comment by StupidRepo** [comment]
<img width="725" alt="image" src="https://github.com/user-attachments/assets/ccc1e8d3-21e0-4520-bfb3-f84666c4c6d6">
@ManlyMarco If you take a look, you can see (to instantiate a networked object), I would need to pass through a prefab name...

**Comment by BadRyuner** [comment]
> <img alt="image" width="725" src="https://private-user-images.githubusercontent.com/69256931/389226037-ccc1e8d3-21e0-4520-bfb3-f84666c4c6d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0MDY4MjgsIm5iZiI6MTczNjQwNjUyOCwicGF0aCI6Ii82OTI1NjkzMS8zODkyMjYwMzctY2NjMWU4ZDMtMjFlMC00NTIwLWJmYjMtZjg0NjY2YzRjNmQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDA3MDg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyNDA5YjljNDhiNmQzOWEzMDc4MmE1N2YyMjQ1MjYxYWE3OTA1NmNkMzg4YmViZDgzZTMxZTllNzZhMzY1NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GzBACrurl4eH4vxD3V1dw7FOlLTpbUTiciASJ2G_TSM"> @ManlyMarco If you take a look, you can see (to instantiate a networked object), I would need to pass through a prefab name. Oh and, also: <img alt="Screenshot 2024-11-23 at 20 47 54" width="884" src="https://private-user-images.githubusercontent.com/69256931/389226130-16d09ad8-e6b1-451b-be5c-00c8305dd660.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0MDY4MjgsIm5iZiI6MTczNjQwNjUyOCwicGF0aCI6Ii82OTI1NjkzMS8zODkyMjYxMzAtMTZkMDlhZDgtZTZiMS00NTFiLWJlNWMtMDBjODMwNWRkNjYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDA3MDg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkMmYxNGVlODg1OTE1MGU5ZGQ2ZGEwMjQxMDIzZDJjYTg4OTU0YjA4Y2MyMDNkZDlkMjA3OTBlNzk2NWJkY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.PA5v6C9JEr_yyywh6tgFe1cMNmY-k-aVqLylW6l1a4E">
> I guess I could make my own RPC instantiation, but why reinvent the wheel?

Use PrefabCache.
1) [Add GameObject to PrerabCache](https://github.com/BadRyuner/PulsarExperiments/blob/master/PulsarExperiments/Prefabs.cs#L72)
2) [Force PrefabCache](https://github.com/BadRyuner/PulsarExperiments/blob/master/PulsarExperiments/Prefabs.cs#L78)

**Comment by StupidRepo** [comment]
Thank you - I did see this after digging around PhotonNetwork class some more. By the way, I would seriously recommend you use [HarmonyPrefix] and [HarmonyPatch(typeof(...))], [HarmonyPatch(nameof(...))]. :)
